0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,671 @@
+
+### Current situation
+
+Currently, the key manager supports two applications: one for generating,
+distributing, and storing master secrets, and the other for ephemeral secrets.
+However, these two applications are not independent, as they both share
+the same key manager policy for secret replication and key derivation.
+
+Issues:
+
+- In the runtime, the logic for key manager status, policy, master secrets,
+  and ephemeral secrets should be decoupled.
+
+- On the host, each application should have its own worker (e.g., a master
+  secret app should have a dedicated worker responsible for participating
+  in the master secret protocol).
+
+### Example apps
+
+Current and future applications:
+
+- Master secrets (for generation and replication master secrets)
+
+- Ephemeral secrets (for generation and replication of ephemeral secrets)
+
+- CPU change (for detecting wether the CPU has changed)
+
+- CHURP (secret sharing scheme)
+
+- Key derivation center (secret sharing scheme)
+
+### App trait
+
+Every application should implement the following trait.
+
+```rust
+/// Key manager application.
+pub trait App {
+  /// Register RPC methods to the dispatcher.
+  ///
+  /// Registered methods are exposed via enclave RPC to the local host
+  /// or remote clients.
+  fn register(&self, ...);
+
+  /// Initialize the application on startup.
+  ///
+  /// Use runtime ID, untrusted local storage, host data, consensus verifier,
+  /// runtime identity, etc. to start the application.
+  fn init(&self, ...);
+}
+```
+
+Each application should register RPC methods, prefixed with the application
+name.
+
+#### Example 1
+
+```rust
+/// Master secrets key manager application.
+pub trait MasterSecrets {
+  fn generate(&self);
+  fn load(&self);
+  fn replicate(&self);
+  fn key_pair(&self);
+  fn private_key(&self);
+  fn public_key(&self);
+  fn symmetric_key(&self);
+  fn update_status(&self);
+}
+```
+
+Methods:
+
+- `master-secrets/generate`
+
+- `master-secrets/load`
+
+- `master-secrets/replicate`
+
+- `master-secrets/key_pair`
+
+- `master-secrets/private_key`
+
+- `master-secrets/public_key`
+
+- `master-secrets/symmetric_key`
+
+- `master-secrets/update_status`
+
+#### Example 2
+
+```rust
+/// CPU change detection key manager application.
+pub trait CPUChangeDetection {
+  fn encrypt(&self);
+  fn decrypt(&self);
+}
+```
+
+Methods:
+
+- `cpu-change/encrypt`
+
+- `cpu-change/decrypt`
+
+### App worker
+
+Each key manager application should have a dedicated worker on the host node
+responsible for communicating with the app and ensuring its consensus view
+is up-to-date.
+
+```go
+// AppWorker is a key manager application worker.
+type AppWorker interface {
+  // Start starts the worker.
+  Start()
+  // Stop stops the worker.
+  Stop()
+}
+```
+
+#### Example
+
+A master secret worker should be responsible for participating in the master
+secret protocol.
+
+## CHURP
+
+CHURP is a proactive secret sharing scheme in which the committee of nodes
```

Comment:
What were the alternative algorithms under consideration? I understand that this one has desirable properties, but this one was also advertised very well. Are there any deployed/reference implementations of CHURP or is this a heavy R&D project? What do incumbents like Lit and Passport do?

Reply:
> What were the alternative algorithms under consideration?

Currently, none. KDC is an option, but there the committee cannot not change, so we would need to replicate secret shares from old enclaves to the new ones.

> Are there any deployed/reference implementations of CHURP or is this a heavy R&D project?

Not that I'm aware of. I'm only aware of Bennet's work and what was in the paper.

> What do incumbents like Lit and Passport do?

Not familiar with these solutions. Can you provide links?

Reply:
https://developer.litprotocol.com/v3/
https://0xpass.gitbook.io

Reply:
I didn't find anything technical from 0xpass, so I can't comment on what scheme they are using.

Correct me if I'm wrong, Lit seems to use a network-wide(?) threshold secret shared key pair (BLS or ECDSA) that is used to threshold-encrypt/decrypt data (without DKG?).

Reply:
I believe you're correct, but I also think they have DKG planned. Right now they're focusing on being a cloud-like KM because that's where most of the value is presently (smart wallets, mainly)

Reply:
CHURP can be used in conjunction w/ KDC to change the committee composition.